pyquil/operator_estimation.py

```diff
@@ -0,0 +1,428 @@
+import itertools
+import json
+import logging
+from json import JSONEncoder
+from math import pi
+from typing import List, Union
+
+import networkx as nx
+import numpy as np
+from dataclasses import dataclass
+from networkx.algorithms.approximation.clique import clique_removal
+
+from pyquil import Program
+from pyquil.api import QuantumComputer
+from pyquil.gates import *
+from pyquil.paulis import PauliTerm
+
+log = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True)
+class Experiment:
+    """
+    One tomography-like experiment.
+
+    Many near-term quantum algorithms take the following form:
+        - Start in a pauli state
+        - Prepare some ansatz
+        - Measure it w.r.t. pauli operators
+
+    Where we typically use a large number of (start, measure) pairs but keep the ansatz preparation
+    program consistent. This class represents the (start, measure) pairs. Typically a large
+    number of these :py:class:`Experiment` objects will be created and grouped into
+    an :py:class:`ExperimentSuite`.
+    """
+    in_operator: PauliTerm
+    out_operator: PauliTerm
+
+    def __str__(self):
+        return f'{self.in_operator}\u2192{self.out_operator}'
+
+    def __repr__(self):
+        return f'Experiment[{self}]'
+
+    def serializable(self):
+        return str(self)
+
+    @classmethod
+    def from_str(cls, s: str):
+        """The opposite of str(exp)"""
+        instr, outstr = s.split('\u2192')
+        return cls(in_operator=PauliTerm.from_str(instr),
+                   out_operator=PauliTerm.from_str(outstr))
```

Comment:
Currently, `from_str(..)` can result in nonsensical things like `Experiment.from_str('1*A0→1*Z0')`, which if you print out, will give `‘1*→1*Z0’`

Reply:
good find. Adding a test and input validation